apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/Numeric.cs

```diff
@@ -452,14 +452,38 @@ public decimal ToDecimal(LossOfPrecisionHandling lossOfPrecisionHandling)
 
             // FIXME: Awful performance!
             string text = ToString();
-            // Handles overflow
+
+            // For non-extreme values, we can just parse the text and we're fine.
+            int maxSafeLength = 28 + (_scaledValue < 0 ? 1 : 0) + (text.Contains('.') ? 1 : 0);
+            if (text.Length < maxSafeLength)
+            {
+                return decimal.Parse(text, CultureInfo.InvariantCulture);
+            }
+
+            // Otherwise, we may need to care about a loss of precision. Try a round-trip first.
             decimal dec = decimal.Parse(text, CultureInfo.InvariantCulture);
-            if (lossOfPrecisionHandling == LossOfPrecisionHandling.Throw && this != Type.ValueFromDecimal(dec, LossOfPrecisionHandling.Truncate))
+            if (this == Type.ValueFromDecimal(dec, LossOfPrecisionHandling.Truncate))
+            {
+                return dec;
+            }
+            // If we've lost precision already and that's unacceptable, throw.
+            if (lossOfPrecisionHandling == LossOfPrecisionHandling.Throw)
             {
-                // TODO: Is this the right exception to use?
                 throw new InvalidOperationException("Conversion would lose information");
             }
-            return dec;
+            // Otherwise, we need to truncate - which means truncating *before* parsing.
+            int length = Math.Min(text.Length - 1, maxSafeLength + 3);
```

Comment:
I may be missing something obvious, but I couldn't understand the +3 to maxSafeLength. (given that sign and '.' are already included in maxSafeLength)

Reply:
It's very much a "this is just a prototype, my maths might be wrong - this is inefficient but safe" :)

But we would at least want `maxSafeLength + 1`, as there might be a digit which *can* be parsed, but isn't safe to *assume* will be parsed just based on the length.

Reply:
Right Jon :) Thanks for helping me understand. 

One more thought - If we plan to use the same fix everywhere, PgNumeric (where we can have very large string due to large precision and scale) may suffer a little with substring in loop.

Reply:
We should definitely apply better than "Friday-afternoon thinking" for the loop - but even if the original string is very large, we'd start off with a relatively small substring due to the `Math.Min` call. We shouldn't need more than two or three iterations.